### PR TITLE
[SYCL][NFCI] Refactor device code split implementation once again

### PR DESCRIPTION
**Pull request by AlexeySachkov**
Apology for a not so small PR (or rather PR description?) in advance.
The PR is marked as `NFCI`, because no functional changes are intended, but I'm not 100% sure if there are corner-cases when behavior changes.

#### Intro

This is a refactoring of how we perform device code split in `sycl-post-link`, which is intended to solve several existing issues with the current implementation:
1. increased peak RAM consumption by `sycl-post-link`
2. bad scaling with more and more split "dimensions" being added
3. increased tests maintenance cost due to non-deterministic order (between commits) of output files produced by `sycl-post-link`

#### A bit more context about the issues above:

(1) Increase peak RAM consumption is caused by the fact that we currently preserve **all** splits in-memory, even though we can process them on-by-one and discard them as soon as we stored them to a disk. This was implemented as a memory consumption optimization in #5021, but it got accidentally reverted in #7302 as an attempt to workaround (2).

(2) is pretty much summarized in our source code:

https://github.com/intel/llvm/blob/afebb2543ccecb89f83c84b68fba7616bbab89ac/llvm/tools/sycl-post-link/sycl-post-link.cpp#L806-L811

(3) is caused by a bad implementation decision made in #7302: because every split is now identified by a hash, every time you add a new split "dimension"/new feature to an account, it results in different hashes for existing tests. Just look how many unrelated tests had to be updated in #7512, #8056 and #8167

#### Now to the PR itself:

It introduces a new infrastructure for categorizing/grouping kernel functions: instead of using hashes, we now build a string description for each kernel function and then group kernels with the same description string together.

String description is built by a new entity: it accepts a set of rules, where each rule is a simple function which returns a string for passed `llvm::Function`. Results of all rules are concatenated together and rules are invoked in a stable order of their registration.

There is a simple API for building those rules. It provides some predefined rules for the most popular use cases like turning a function attribute or a metadata into a string descriptor for the function. There is also a possibility to pass a custom callback there to implement more complicated logic.

#### How does this PR help with issues above?

(1) and (2) are fixed in conjunction: `sycl-post-link` was refactored to avoid storing more than one split module at a time and that is possible because the PR unifies per-scope and optional-kernel-features splitters into a single generic splitter. The new API for kernels categorization seems to be flexible enough to provide that infrastructure so merged splitters still look OK code-wise.

(3) is caused by using string identifiers instead of hashes as well as by using a data structure which sorts identifiers.

#### Any other benefits from this PR?

About 50 lines of code less to support :)

Extending device code split for more optional features would be even easier than it is now: instead of adding several changes to various places around `UsedOptionalFeatures` structure, it will be just adding a 1-3 lines of code. Please also note that `UsedOptionalFeatures` contains tons of inconsistencies in its implementation, which will all gone with this PR: in `operator==` we don't use hash and instead compare certain fields directly (and we do miss some of them); `generateModuleName` method skips some of optional features and ignores them.

Cross-module `device_global` usages checks should now work at all split dimensions (except for ESIMD).

#### Any potential downsides?

With current `UsedOptionalFeatures` there is a possibility to embed various information (used aspects, `large-grf` flag, etc.) directly during device code split to avoid re-gathering that information later when we generate properties. With the suggested approach, it would be harder to do, because it doesn't seem to naturally fit to the proposed infrastructure: see changes I did around `large-grf` in this PR.

However, we have never actually implemented this and re-querying some metadata from function doesn't seem like a bottleneck, so it should really be a very minor and only theoretical downside.

### CODE REVIEW

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (sarnex) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:572]: do we even need the esimd suffix for anything? the runtime actually checks the properties, so this doesn't actually affect behavior right? i would not be against removing this completely.

**Review by sarnex — APPROVED** [review]
wow, doing it this way is so much simpler and easier to understand. there's so much less nonsense now. thanks for doing this!

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:572]: The only difference this make is names of temporary filenames we produce. Could be useful in tests and/or debugging, but has no real impact for users (except for maybe contributing towards encountering an edge case of extending path size limitation on windows?)

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: Done in 2a367012d3ea0b86fe5b41431ca1a1bddf959875

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (bader) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:464]: I suggest using `std::find_if` instead of manually written loop.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (bader) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:877]:
```suggestion
  if (DoSplit)
    return std::make_unique<ModuleSplitter>(std::move(MD), std::move(Groups));

  return std::make_unique<ModuleCopier>(std::move(MD), std::move(Groups));
```

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Switched to `llvm::any_of` in aaf1e92b82415656faec197d9afa340fe54627bd

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: Thanks, applied in aaf1e92b82415656faec197d9afa340fe54627bd

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (asudarsa) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp:228]: AFAIU, This attribute can also be provided by the user via the SYCL_EXTERNAL keyword? Thanks

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (asudarsa) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:697]: Please add a comment on what these kinds stand for. Thanks

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (asudarsa) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:740]: Do we want to add a delimiter in the result string between rules? One corner case is when two attributes have values like this kernel1: 1 and 20; kernel2: 12 and 0. When we concatenate string for both cases we get 120. 

Thanks

**Review by asudarsa — COMMENTED** [review]
Hi @AlexeySachkov 

Overall it looks good to me. I have a couple of comments that can benefit from addressing. I would like to take a second look by end of day today to look for corner cases and then approve. Hope that's agreeable.

It will also be a good idea to have a few lines explaining the design, either in the form of comments at appropriate location or a short design document.

Thanks again for a great refactoring.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp:228]: > AFAIU, This attribute can also be provided by the user via the SYCL_EXTERNAL keyword? Thanks

That's true, but I don't see any issues with it still: `isSYCLExternalFunction` checks if `sycl-module-id` attribute is present and `isEntryPoints` checks the same for non-kernels. Therefore, this check only works for kernels, which are implicitly handled by the compiler.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp:697]: Added brief descriptions in 13169d46b86d40196060de52bfe1dff86d75409f

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp:746]: Good point, thanks! Added a delimiters in 67082dda9a95d3dabaa3c4c9158e86f8b1b5ba27

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (asudarsa) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:754]: Do we need a delimiter here as well? Thanks

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks Alexey.

**Comment by AlexeySachkov** [comment]
@sarnex, @asudarsa, I pushed one more commit (cf3a36576a91129a36b8f3086bfc7f063855a815) with minor changes like: new comments, removed `default` case, moved some code to anonymous namespace. I assume that the change is small and trivial so it doesn't require re-approval. I plan to merge the patch when pre-commit finished. If there is some feedback, please let me know, I will apply it pre/post commit.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (sarnex) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:693]: nit: could we make a type alias for what the tuple represents? tuple<stringref * 3> is not super intuitive to me

**Review by sarnex — APPROVED** [review]
lgtm feel free to address later

**Comment by sarnex** [comment]
@AlexeySachkov Hey, are there any plans to merge this soon? I'm planning to extend the large grf support we have in SYCL and that will interact with splitting, so I'd be cool to get this change in.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: Applied in 7ff7531215698fee717b3dadcb0987b3af1dee27. Not sure about the type alias name, please let me know if there are better suggestions

**Comment by AlexeySachkov** [comment]
@sarnex, @asudarsa, sorry for delay. I've rebased the PR on top of #8763 and it is now ready for review. Changes since last update:
- 863abed65c5951fff893438947177befa43cf156 fixed a warning about improper `sycl-post-link` options being emitted even for valid cases
- b80c1c40b5cd2df0b6e9c377fcb2a109f5a157a5 is a merge commit. It is buildable, but LIT tests fail, because it essentially reverts all modification to module splitters which were done in #8763
- adadb0d7e99b322530c9643ac33da8827cd628a4 restores reverted functionality. It is a good example, which highlights:
** simplicity of extending device code split (just compare amount of changes made in `ModuleSplitter.cpp` with #8763);
** inability to propagate properties computed at device code split phase to `sycl-post-link`: we don't really "compute" them anymore. That's the same thing as with `large-grf` which is mentioned in PR description
** amount of changes in tests after extending a new splitter is now minimal: the order of output modules is stable again
- 9158ca89088e7508c81b9f56e79f2bebf1e91dbd fixes incorrect merge in one of tests
- 7ff7531215698fee717b3dadcb0987b3af1dee27 fixes comment from @sarnex 

I would like you to take another look at the PR before I merge it, to review recent changes

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: Ok, I think I didn't get your original intent at first. `TupleOfThreeStringRef` is the same as `std::tuple<StringRef * 3>`, it does not represent domain-specific meaning of that data structure.

What about replacing the tuple with the following `struct` instead:
```
struct FlagRuleData {
  StringRef Name, IfPresentStr, IfAbsentStr;
};
```

This way it actually has some meaning and will allow to get rid of `std::get<N>`, which are not intuitive either. Instead we would access fields through meaningful names

Comment (sarnex) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: yeah that struct looks really good, thx

Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: Implemented that `struct` approach in e4b452e6142f5d1175758213a42f9821740dc85e

**Review by sarnex — APPROVED** [review]
looks great to me, only nits, thanks a lot! looking forward to making use of this soon!

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: Applied in d61e93e7fac6c82f3bfdfcb5ede68ea5d96d555c

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: Thanks! Applied in d61e93e7fac6c82f3bfdfcb5ede68ea5d96d555c

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: Applied in d61e93e7fac6c82f3bfdfcb5ede68ea5d96d555c

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: Applied in d61e93e7fac6c82f3bfdfcb5ede68ea5d96d555c

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: That understanding is correct. Added a comment explaining this in d61e93e7fac6c82f3bfdfcb5ede68ea5d96d555c

**Review by sarnex — APPROVED** [review]

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp
Comment (asudarsa) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:856]: Looks good. Thanks

**Review by asudarsa — APPROVED** [review]
Hi @AlexeySachkov 

Overall looks good to me. Thanks

**Review by sarnex — APPROVED** [review]
new changes lgtm also, thanks.

in my experience invoke_simd is very sensitive to the environment, so im not surprised changing the optlevel causes an issue. dropping the flag and making a bug for the gpu people makes sense, ill email you who to assign it to

**Comment by AlexeySachkov** [comment]
Merge with `sycl` branch to properly restart CI: since I had regressions on L0 at some point, I don't want to merge this without making sure that I actually fixed them

**Pull request merged**
